skopt/space.py

```diff
@@ -0,0 +1,458 @@
+
+    @property
+    def transformed_size(self):
+        return len(self.categories)
+
+    @property
+    def bounds(self):
+        return self.categories
+
+    @property
+    def transformed_bounds(self):
+        return [(0.0, 1.0) for i in range(len(self.categories))]
+
+
+class Space:
+    """Search space."""
+
+    def __init__(self, dimensions):
+        """Initialize a search space from given specifications.
+
+        Parameters
+        ----------
+        * `dimensions` [list, shape=(n_dims,)]:
+            List of search space dimensions.
+            Each search dimension can be defined either as
+
+            - a `(upper_bound, lower_bound)` tuple (for `Real` or `Integer`
+              dimensions),
+            - a `(upper_bound, lower_bound, "prior")` tuple (for `Real`
+              dimensions),
+            - an instance of a `Dimension` object (for `Real`, `Integer` or
+              `Categorical` dimensions), or
+            - as a list of categories (for `Categorical` dimensions).
+        """
+        _dimensions = []
+
+        for dim in dimensions:
+            if isinstance(dim, Dimension):
+                _dimensions.append(dim)
+
+            elif (len(dim) == 3 and
+                  isinstance(dim[0], numbers.Real) and
+                  isinstance(dim[2], str)):
+                _dimensions.append(Real(*dim))
+
+            elif len(dim) > 2 or isinstance(dim[0], str):
+                _dimensions.append(Categorical(*dim))
+
+            elif isinstance(dim[0], numbers.Integral):
+                _dimensions.append(Integer(*dim))
+
+            elif isinstance(dim[0], numbers.Real):
+                _dimensions.append(Real(*dim))
+
+            else:
+                raise ValueError("Invalid grid component (got %s)." % dim)
+
+        self.dimensions = _dimensions
+
+    def rvs(self, n_samples=1, random_state=None):
+        """Draw random samples.
+
+        Parameters
+        ----------
+        * `n_samples` [int, default=1]:
+            Number of samples to be drawn from the space.
+
+        * `random_state` [int, RandomState instance, or None (default)]:
+            Set random state to something other than None for reproducible
+            results.
+
+        Returns
+        -------
+        * `points`: [array-like, shape=(n_points, n_dims)]
+           Points sampled from the space.
+        """
+        rng = check_random_state(random_state)
+
+        # Draw
+        columns = []
+
+        for dim in self.dimensions:
+            if sp_version < (0, 16):
+                columns.append(dim.rvs(n_samples=n_samples))
+            else:
+                columns.append(dim.rvs(n_samples=n_samples, random_state=rng))
+
+        # Transpose
+        rows = []
+
+        for i in range(n_samples):
+            r = []
+            for j in range(self.n_dims):
+                r.append(columns[j][i])
+
+            rows.append(r)
+
+        return np.asarray(rows)
```

Comment:
Yes, I agree. I looked at this file first and hence assumed that the `_optimize` functions had been refactored to accept categorical parameters.